### ISSUE TITLE
Evaluating a saved model on another Test dataset

### ISSUE THREAD

**Issue by AhmedIdr** [deepset-ai/FARM#77]
Hey guys, 

i was just wondering if there is a way to evaluate a saved model (maybe I missed it in the docs). I managed to do that using the Inferencer but it is quite slow and it doesn't seem like the right way to it. Thank you in advance.

**Comment by Timoeller** [comment]
Hey @AhmedIdr, that is of course a bummer. Lets try to fix the issue on your side or we make the inferencer more fast. But first I need some more information to answer your question.
With saved model you mean 
1. a model that was trained on a downstream task that you want to evaluate on a similar or the same downstream task or
2. a raw language model that was either pretrained from scratch or adjusted to domain language and you want to evaluate on a bunch of downstream tasks?

Furthermore can you tell which part of the inferencer isn't working smoothly? Is it the actual inference using computations inside the NN and do you have a GPU + GPU enabled? Or is it the eval data transformation that takes long?
Thanks

**Comment by AhmedIdr** [comment]
Hey @Timoeller, 
With saved model i meant a model that was trained on a my downstream task and i want to evaluate on the same task but with another test dataset. 
I basically just the doc_classification example and used the inferencer just like in the example but on a whole test dataset and then i extracted the predicted labels from the results and used sklearn to get f1-score. And yes im using a GPU. This doesn’t seem like the right way to do it and just like a hacky way to get the results. 
Kind regards
Ahmed

**Comment by Timoeller** [comment]
Ok, very true, using the Inferencer is not really suited for this use-case.

The quick fix is to retrain your doc_classification model and just pass a different test_dataset String to the TextClassificationProcessor. (Though you cannot reuse your previously trained model.)

If you want to reuse your previously trained model you would need to stick together the parts yourself. It should be rather easy, very similar to code in the doc_classification example.
You will need 
1. DataSilo that correctly loads your other test dataset (needs corresponding processor)
2. AdaptiveModel that loads the previously trained and saved model
3. Evaluator that gets passed the DataSilo from 1. and then calls eval() with the model loaded at 2. You can have a look in train.py to see how the Evaluator is instantiated and used.

This should do it. Please give me a shout when you run into problems or need more clarification. When you are done it would be super nice if you could paste your code in this issue, so other people can reuse it. We might refactor your code for a tutorial or even include it in the codebase when more people need this functionality.

**Comment by AhmedIdr** [comment]
Okay, i ll try to take a look at this later. If I managed to solve the problem like you described, I will post my code.

**Comment by Timoeller** [comment]
@tanaysoni will take over from here. We found a clean solution for the Inferencer to work on datasets.

**Comment by AhmedIdr** [comment]
Sorry I haven't had the time to take care of this. I just used my workaround using the Inferencer. This is how I used it. 
`
        for i in test[0]:
           info = {}
           info['text'] = i
           mydict.append(info)

        model = Inferencer.load(save_dir,batch_size=len(mydict))
        result = model.run_inference(dicts=mydict)
        y_pred = []
        for i in result[0].get('predictions'):
            y_pred.append(i.get('label'))
        f1.append(f1_score(actu,y_pred, average=average_method))
`

**Comment by Timoeller** [comment]
We implemented a function to do [inference from files](https://github.com/deepset-ai/FARM/blob/master/farm/infer.py#L211) that wraps this functionality into FARM. 

Labels are not propagated and need to be added to the predictions. Concerning the order of predictions: Disable multiprocessing (by setting max_processes = 1 in the inference_from_file() function) to make sure to have the same order in predictions and labels.

Closing this now.
